### PR TITLE
Fix issues with onchange on relational 2many fields

### PR DESCRIPTION
**Pull request by rco-odoo**
Use the method `convert_to_onchange` to convert cached values in the output format of method `onchange`. The format of `one2many` and `many2many` fields is modified to return the values of the fields instead of a diff with commands.

### CODE REVIEW

**Comment by Whisno** [comment]
Where's the thumbs up button ?

File: addons/web/static/src/js/views/form_relational_widgets.js [hunk 1]
```diff
@@ -471,10 +471,22 @@ var AbstractManyField = common.AbstractField.extend({
...
         }
+        if (_.find(ids, function(id) { return id instanceof Object; } )) {
+            return this.send_commands(_.map(ids, function (obj) {
+                if (obj.id) {
+                    return COMMANDS.update(obj.id, obj);
+                } else {
+                    return COMMANDS.create(obj);
+                }
+            }));
+        }
```
Comment (rco-odoo) [inline addons/web/static/src/js/views/form_relational_widgets.js]: Be careful: arrays are instances of `Object`! So if you invoke `set_value` with commands, you will execute both `if` statements, and the second one will probably crash.

I suggest to replace both `if` statements above by something like:

    return this.send_commands(_.map(ids, function (command) {
        if (command instanceof Array) {
            return command;
        } else if (command.id) {
            return COMMANDS.update(command.id, command);
        } else {
            return COMMANDS.create(command);
        }
    }));

File: addons/web/static/src/js/framework/data.js [hunk 2]
```diff
@@ -737,43 +739,123 @@ var BufferedDataSet = DataSetStatic.extend({
             self.last_default_get = res;
         });
     },
+    _update_cache: function (id, options) {
+        var cached = _.detect(this.cache, function(x) {return x.id === id;});
```
Comment (Gorash) [inline addons/web/static/src/js/framework/data.js]: one should call this method after modifying `this.from_read`, `this.to_create` or `this.to_write`, and updates `this.cache` and `this.from_readonly`

File: addons/web/static/src/js/framework/data.js [hunk 3]
```diff
@@ -737,43 +739,123 @@ var BufferedDataSet = DataSetStatic.extend({
...
+                    _.each(data, function (v, k) {
+                        if (!_.isEqual(v, cached.values[k])) {
+                            record_to_create.values[k] = v;
+                            dirty = true;
+                        } else {
+                            delete data[k];
+                        }
+                    });
+                } else {
+                    // check if the record change
```
Comment (Gorash) [inline addons/web/static/src/js/framework/data.js]: ... and update `to_write` accordingly (possibly remove it)

File: addons/web/static/src/js/framework/data.js [hunk 4]
```diff
@@ -803,21 +890,20 @@ var BufferedDataSet = DataSetStatic.extend({
     read_ids: function (ids, fields, options) {
```
Comment (Gorash) [inline addons/web/static/src/js/framework/data.js]: read whatever is needed from the server to have ids in this.from_read

File: addons/web/static/src/js/views/form_relational_widgets.js [hunk 5]
```diff
@@ -472,11 +474,11 @@ var AbstractManyField = common.AbstractField.extend({
             throw new Error("set_value of '"+this.name+"' must receive an list of ids without virtual ids.", ids);
         }
         if (_.find(ids, function(id) { return typeof(id) !== "number"; } )) {
-            this.dataset.reset_ids([]);
-            return this.send_commands(ids);
+            this.dataset.reset_ids([], {keep_read_data: true});
+            return this.send_commands(ids, {'_inhibit_on_change_flag': this._inhibit_on_change_flag});
```
Comment (Gorash) [inline addons/web/static/src/js/views/form_relational_widgets.js]: here: convert ids dicts to commands

File: addons/web/static/src/js/views/form_relational_widgets.js [hunk 6]
```diff
@@ -576,6 +578,13 @@ var AbstractManyField = common.AbstractField.extend({
 
         _.each(command_list, function(command) {
             self.mutex.exec(function() {
+                if (!Array.isArray(command)) {
```
Comment (Gorash) [inline addons/web/static/src/js/views/form_relational_widgets.js]: use `instanceof Array` instead

File: addons/web/static/src/js/views/form_relational_widgets.js [hunk 7]
```diff
@@ -879,9 +900,9 @@ var X2ManyDataSet = data.BufferedDataSet.extend({
         var self = this;
         var def = this._super(data, options);
         def.then(function (id) {
-            setTimeout(function () {
+            _.defer(function () {
```
Comment (Gorash) [inline addons/web/static/src/js/views/form_relational_widgets.js]: remove this one

File: addons/web/static/test/x2many.js [hunk 8]
```diff
@@ -240,6 +240,44 @@ Tour.register({
...
+        },
+        {
+            title:      "test one2many triggered the onchange on save for the line",
+            waitFor:    '.oe_list_content td.oe_list_field_cell.oe_readonly:contains([test_trigger] Demo User:e)',
+        },
+        {
+            title:      "test one2many triggered the onchange on save",
+            waitFor:    'textarea[name="message_concat"]:propValueContains([test_trigger] Demo User:e)',
+        },
+*/
```
Comment (Gorash) [inline addons/web/static/test/x2many.js]: move the tour on a new branch

**Pull request merged**